### ISSUE TITLE
curation

### ISSUE THREAD

**Issue by torkus** [layday/github-wow-addon-catalogue#1]
This is a cool repo - I've been thinking for a long time about how to gather these addons together. There are also a few on gitlab as well.

The sticking point for me was always the assumption that someone was curating and vetting these addons. It would be awful if we became vectors for malicious addons.

How does one get on the list? How does one get an addon taken off the list? Or whitelisted/blacklisted? If blacklisted, why blacklisted? Has a virus scanner been used (does Github also do that ...?).

Anyway, great work pulling this data together.

**Comment by layday** [comment]
All good questions. I haven’t thought that far if I’m honest. But if we concede that some of these we cannot reasonably fix, then we need to be able to evaluate their severity and warn the user appropriately.

If something more polished should come out of this, I’d want to break up the CSV and accept PRs for additions/deletions. A service would run to update addon metadata on a schedule and a bot could open PRs for newly-harvested addons.

We’d also need to put some processes and criteria in place for when we might need to delist or refuse to list an addon.

Sent with [ProtonMail](https://protonmail.com/) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by torkus** [comment]
https://github.com/layday/github-wow-addon-catalogue/blob/main/collect.py#L65

    (API_URL / "repos" / repo["full_name"] / "releases").with_query(per_page=1)

I was taught operator overloading in C++ in the early 00s and knew then, when I knew nothing about everything, that it was a curse. I hope it's just `yarl` being clever and not a sign of the direction the language is taking.

I hadn't come across `@dataclass(frozen=True)` yet (3.8+ it seems) but this seems like a write-once named tuple which ain't so bad. I implemented something similar about a year ago but was borrowing ideas of immutability from Clojure at the time.

`async` in both Python and Clojure seems ham-fisted, but at least in Clojure it's just a library and didn't have to be integrated into the language.

f-strings seem like a good idea and I enjoy using them. For some reason they remind me of PHP's early use of global variables for everything. It will be good to eventually support them when 3.6 expires and we all (the LTS crowd) move on to 3.8.

https://github.com/layday/github-wow-addon-catalogue/blob/main/collect.py#L166

You know, if it works it works ;) I bet you're not the only bot hammering away at the Github API. A breather in between requests is just common courtesy.

I'm procrastinating, I have to make a design decision in strongbox around release.json files and I'd thought I'd just go through how you're building up your list.

> But if we concede that some of these we cannot reasonably fix, then we need to be able to evaluate their severity and warn the user appropriately.

I think it should err on the side of inclusiveness. Capture every addon it can, mark those not using `release.json` as such and keep a separate 'reputation' list for any egregiously bad behaviour that might affect end users. It's a trade-off between being a central authority/gatekeeper and impartiality. You can exercise more control and your best judgement with the smaller reputation list. If someone complains an addon stole their code and changed the licence etc, you don't need to take it off the catalogue, but you can put a strike against it on the shit list. Or the fork may actually be legitimate and the licence was changed from MIT/BSD to GPL (totally legal) and is actually better than the original and, while you don't want to get involved in a social brouhaha, you also want to be fair to both parties so your gatekeeper authority isn't brought into disrepair.

Anyway, I should get back to it. Let me know if I can be useful.

**Comment by layday** [comment]
> I was taught operator overloading in C++ in the early 00s and knew then, when I knew nothing about everything, that it was a curse. I hope it's just `yarl` being clever and not a sign of the direction the language is taking.

yarl borrows directly from pathlib which is a module in the stdlib. pathlib overloads the div operator as a platform-agnostic directory separator.  URLs have only got the one separator, so uh... yeah.

dataclasses were added in 3.7, and f-strings in 3.6 so you can start using them now ;)

> Capture evapturery addon it can, mark those not using `release.json` as such and

I suppose, but installing add-ons without a `release.json` is nasty work.  What you're saying is we should reframe the catalogue to be more, let's say, 'exploratory', which makes sense if...

> It's a trade-off between being a central authority/gatekeeper and impartiality.

I like this approach.  In essence this would function as a 'meta-catalogue' and add-on managers would be encouraged to apply their own judgement/criteria.  We should provide enough metadata to help them make that call.

Feel free to open a PR for a new `release.json` col or anything else that you think might need changing.

**Comment by lsh-0** [comment]
> yarl borrows directly from pathlib which is a module in the stdlib. pathlib overloads the div operator as a platform-agnostic directory separator. URLs have only got the one separator, so uh... yeah.

nuts! 

> I suppose, but installing add-ons without a release.json is nasty work. 

eh, not so bad, but explicit is better than implicit.

> Feel free to open a PR for a new release.json col or anything else that you think might need changing.

I have a few things to do before this month's release, then I'm busy figuring out how to gently disable curseforge and migrate users to alternatives until it's turned off for good in Feb. There are ~500 addons in this new catalogue and roughly ~5290 addons on Curserforge that have seen an update since BfA was released. I'm curious where those other 4.7k actively developed addons live if not github/gitlab/bitbucket. Does Curseforge also provide a git forge?

**Comment by torkus** [comment]
(bloody hell. Delete that comment for me please?)

---

> yarl borrows directly from pathlib which is a module in the stdlib. pathlib overloads the div operator as a platform-agnostic directory separator. URLs have only got the one separator, so uh... yeah.

nuts!

> suppose, but installing add-ons without a release.json is nasty work.

eh, not so bad, but explicit is better than implicit.

> Feel free to open a PR for a new release.json col or anything else that you think might need changing.

I have a few things to do before this month's release, then I'm busy figuring out how to gently disable curseforge and migrate users to alternatives until it's turned off for good in Feb. There are ~500 addons in this new catalogue and roughly ~5290 addons on Curserforge that have seen an update since BfA was released. I'm curious where those other 4.7k actively developed addons live if not github/gitlab/bitbucket. Does Curseforge also provide a git forge?

**Comment by layday** [comment]
Yeah, CF has/had its own svn/git/hg service.  And some authors might not version things remotely, or at all.

**Comment by torkus** [comment]
I'm going to fork this repo and use it to generate a Github catalogue for Strongbox. No firm ideas on how to keep it maintained or curate it but just something to support before the end of Dec.

If you have any plans I'd love to hear them. If not I might spin off a mini website where addon repos can be nominated by the public.

**Comment by layday** [comment]
I haven’t had time to work on this but I’d be interested in working on a site like that together. Would also be happy to accept any improvements. If you’d like you can have write permissions here.

Sent with [ProtonMail](https://protonmail.com/) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by torkus** [comment]
I probably won't be modifying anything in the immediate future, it's just a fork to keep the ground from moving under me.

I'll let you know when I have some breathing room to look at this again. Probably Jan.

**Comment by torkus** [comment]
your list of github addons has been incorporated into the strongbox catalogues: https://github.com/ogri-la/strongbox-catalogue/commit/62632ddc0ec4040b14e2a0d7173ff6c978fa3198

Strongbox users can install them like regular addons now. Thanks again for the list, it's been very convenient.

**Comment by layday** [comment]
Sweet. I should get on with it too :)

Sent with [ProtonMail](https://protonmail.com/) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by layday** [comment]
I actually meant to do that last week but I got sidetracked, adding [GitHub auth flow](https://github.com/layday/instawow/blob/94c3ef9f1a26d660a04da64e5f15d834c9e24f6c/src/instawow/cli.py#L849) to the CLI and revamping the configuration.

**Comment by torkus** [comment]
> I actually meant to do that last week but I got sidetracked, adding GitHub auth flow to the CLI and revamping the configuration.

Strongbox supports a `GITHUB_TOKEN` envvar but that's a bit esoteric. I might steal this idea for later next year. The 60 API requests/hr limit (or whatever) is pretty small when I have to inspect the releases page for each addon.

I was wondering though, have you seen a `X-Github-ID` type `.toc` field in the wild?

**Comment by layday** [comment]
I've added support for add-ons without `release.json` in https://github.com/layday/github-wow-addon-catalogue/commit/5b93d9618d45ef94d9f19f1b62dc23cedae6f2ae.

**Comment by torkus** [comment]
very cool - 742 rows vs 540 rows. Github addons are becoming a force to be reckoned with.

I've been on holidays and just got back. I might rig something up to see how many of these can be installed by strongbox. 

My latest release took so long because of code like this: https://github.com/layday/github-wow-addon-catalogue/commit/5b93d9618d45ef94d9f19f1b62dc23cedae6f2ae#diff-30e572b4070dd48a29c4c3c8cddfb3a56d5c9e7b504164d51296a705b5d13bb7R110-R116

I wasn't happy with it defaulting to 'retail' unless [a whole bunch of other checks happened first](https://github.com/ogri-la/strongbox/blob/develop/src/strongbox/github_api.clj#L73-L133), but as I mentioned elsewhere - the cure for ambiguity is the `release.json` file.

Also, should `-tbc` be supported?

**Comment by layday** [comment]
The BigWigs packager doesn't generate `-tbc` filenames, have you seen any of those in the wild?

> My latest release took so long because of code like this: [5b93d96#diff-30e572b4070dd48a29c4c3c8cddfb3a56d5c9e7b504164d51296a705b5d13bb7R110-R116](https://github.com/layday/github-wow-addon-catalogue/commit/5b93d9618d45ef94d9f19f1b62dc23cedae6f2ae#diff-30e572b4070dd48a29c4c3c8cddfb3a56d5c9e7b504164d51296a705b5d13bb7R110-R116)

Yeah, I'm not happy about that either. I suppose we could be downloading the zips and extracting the version from the `toc` file. But unless add-on managers do that too they won't be able to reconcile release archives with game versions from the catalogue.

**Comment by lsh-0** [comment]
> The BigWigs packager doesn't generate -tbc filenames, have you seen any of those in the wild?

yup, I'll see if I can find some examples. `tbc` was more widely used until BigWigs settled on `bcc`.

> Yeah, I'm not happy about that either. I suppose we could be downloading the zips and extracting the version from the toc file. But unless add-on managers do that too they won't be able to reconcile release archives with game versions from the catalogue.

Well, I think the fewer number of API calls the addon managers have to make the better. *Most* users won't be authenticated and will have the 60req/hr cap by default. Then again, downloading and inspecting zip files seems unreasonable.

Strongbox does a broad sweep the first time a github/gitlab addon is 'imported' and this includes downloading and parsing any `.toc` files in the root of the repository looking for `## Interface` values. The game tracks found in this phase are used to fill in any gaps later.

**Comment by torkus** [comment]
> The BigWigs packager doesn't generate -tbc filenames, have you seen any of those in the wild?

yup, I'll see if I can find some examples. `tbc` was more widely used until BigWigs settled on `bcc`.

> Yeah, I'm not happy about that either. I suppose we could be downloading the zips and extracting the version from the toc file. But unless add-on managers do that too they won't be able to reconcile release archives with game versions from the catalogue.

Well, I think the fewer number of API calls the addon managers have to make the better. *Most* users won't be authenticated and will have the 60req/hr cap by default. Then again, downloading and inspecting zip files seems unreasonable.

Strongbox does a broad sweep the first time a github/gitlab addon is 'imported' and this includes downloading and parsing any `.toc` files in the root of the repository looking for `## Interface` values. The game tracks found in this phase are used to fill in any gaps later.
